From 75c2017f8505ef15697b0b917c4c8c6dacbeb01e Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Fri, 11 Dec 2015 18:17:48 -0800
Subject: [PATCH 659/827] drm: xilinx: dp: Enable downspread when sink supports

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

Check DPCD and if sink supports downspread, enable downspread in
the controller register and the sink through aux.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
(cherry picked from commit 5dab41c3eb8acd273f26f36b5598cbaa0703f2c0)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/gpu/drm/xilinx/xilinx_drm_dp.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/gpu/drm/xilinx/xilinx_drm_dp.c b/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
index ece15f6..0e201d2 100644
--- a/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
+++ b/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
@@ -38,6 +38,7 @@
 #define XILINX_DP_TX_ENHANCED_FRAME_EN			0x8
 #define XILINX_DP_TX_TRAINING_PATTERN_SET		0xc
 #define XILINX_DP_TX_SCRAMBLING_DISABLE			0x14
+#define XILINX_DP_TX_DOWNSPREAD_CTL			0x18
 #define XILINX_DP_TX_SW_RESET				0x1c
 #define XILINX_DP_TX_SW_RESET_STREAM1			(1 << 0)
 #define XILINX_DP_TX_SW_RESET_STREAM2			(1 << 1)
@@ -669,6 +670,15 @@ static int xilinx_drm_dp_train(struct xilinx_drm_dp *dp)
 		aux_lane_cnt |= DP_LANE_COUNT_ENHANCED_FRAME_EN;
 	}
 
+	if (dp->dpcd[3] & 0x1) {
+		xilinx_drm_writel(dp->iomem, XILINX_DP_TX_DOWNSPREAD_CTL, 1);
+		drm_dp_dpcd_writeb(&dp->aux, DP_DOWNSPREAD_CTRL,
+				   DP_SPREAD_AMP_0_5);
+	} else {
+		xilinx_drm_writel(dp->iomem, XILINX_DP_TX_DOWNSPREAD_CTL, 0);
+		drm_dp_dpcd_writeb(&dp->aux, DP_DOWNSPREAD_CTRL, 0);
+	}
+
 	ret = drm_dp_dpcd_writeb(&dp->aux, DP_LANE_COUNT_SET, aux_lane_cnt);
 	if (ret < 0) {
 		DRM_ERROR("failed to set lane count\n");
-- 
2.9.3

